docs/recipes/endpoint-testing-with-mongoose.md

```diff
@@ -0,0 +1,281 @@
+
+This recipe shows you how to test your endpoints with AVA and Mongoose.
+
+## Setup
+
+This recipe uses the following libraries:
+
+1. [MongoDB memory server](https://github.com/nodkz/mongodb-memory-server) (A MongoDB in-memory Server)
+2. [Babel Polyfill](https://babeljs.io/docs/usage/polyfill/) (required for MongoDB memory server)
+3. [Supertest](https://github.com/visionmedia/supertest) (An endpoint testing library)
+4. Mongoose
+
+Install the first three libraries by running the following code:
+
+```console
+$ npm install --save-dev mongodb-memory-server babel-polyfill supertest
+```
+
+You should have Mongoose installed already. If not, run the following code to install it:
+
+(Note: You need at least mongoose v4.11.3)
+
+```console
+$ npm install mongoose
+```
+
+## Setting up AVA
+
+Since MongoDB Memory server requires Babel polyfill to work, the easiest way to set up AVA is through the `ava` key in your `package.json` file.
+
+```json
+"ava": {
+    "require": [
+      "babel-register",
+      "babel-polyfill"
+    ]
+  },
+```
+
+If you want AVA to run your tests whenever changes are made, add the `--watch` option.
+
+```json
+"scripts": {
+  "watch": "ava --watch"
+},
+```
+
+## Your test file
+
+First, include the libraries you need.
+
+```js
+// Libraries required for testing
+import test from 'ava'
+import request from 'supertest'
+import MongodbMemoryServer from 'mongodb-memory-server'
+import mongoose from 'mongoose'
+
+// Your server and models
+import app from '../server'
+import User from '../models/User'
+```
+
+### Your server file
+
+In your `server` file, you cannot use `app.listen` to start your app. Export it to another file and call `app.listen` there.
+
+This `server` file should resemble the following:
+
+```js
+const express = require('express')
+const bodyParser = require('body-parser')
+const routes = require('./routes')
+const app = express()
+
+// ======================================
+// # Middlewares
+// ======================================
+app.use(bodyParser.json())
+
+// ======================================
+// # Routes
+// ======================================
+app.get('/litmus', async (req, res) => {
+  const { email } = req.body
+  res.json(await User.findOne({email}))
+})
+
+app.post('/litmus', async (req, res) => {
+  const { email, name } = req.body
+  const user = new User({email, name})
+  res.json(await user.save())
+})
+
+module.exports = app
+```
+
+### Back to your test file
+
+**First, start your MongoDB instance and connect to Mongoose:**
+
+```js
+// Start MongoDB Instance
+const mongod = new MongodbMemoryServer()
+
+// Create connection to mongoose before all tests
+test.before(async t => mongoose.connect(await mongod.getConnectionString(), { useMongoClient: true }))
+```
+
+When you run your first test, MongoDB downloads the latest MongoDB Binaries. It may take a minute. (The download is ~70mb).
+
+Note: Since we're using async/await, you need Node v7.6 and above.
```

Comment:
AVA takes care of this for you. I'd assume (especially when `regenerator-runtime` is loaded this recipe works all the way back on Node.js 4).